plugins/expressions-converter/src/org/jetbrains/kotlinx/dataframe/ExplainerIrTransformer.kt

```diff
@@ -0,0 +1,269 @@
+        if (expression.startOffset < 0) return expression
+        if (expression.type.classFqName in dataFrameLike) {
+            return transformDataFrameExpression(expression, expression.symbol.owner.name, receiver = null, data)
+        }
+        return super.visitExpression(expression, data)
+    }
+
+    // also, what if expression type is not DataFrame, but Unit? and receiver expression is DataFrame at some point
+    override fun visitCall(expression: IrCall, data: ContainingDeclarations): IrExpression {
+        if (expression.startOffset < 0) return expression
+        if (expression.type.classFqName in dataFrameLike) {
+            if (expression.symbol.owner.name == Name.identifier("component1")) return expression
+            var receiver = expression.extensionReceiver
+            // expression.extensionReceiver = extension callables,
+            // expression.dispatchReceiver = member callables such as "GroupBy.aggregate"
+            if (receiver != null) {
+                val transformedExtensionReceiver = expression.extensionReceiver?.transform(this, data)
+                expression.extensionReceiver = transformedExtensionReceiver
+            } else {
+                receiver = expression.dispatchReceiver
+                val transformedExtensionReceiver = expression.dispatchReceiver?.transform(this, data)
+                expression.dispatchReceiver = transformedExtensionReceiver
+            }
+
+            return transformDataFrameExpression(expression, expression.symbol.owner.name, receiver = receiver, data)
+        }
+        return super.visitExpression(expression, data)
+    }
+
+    @OptIn(FirIncompatiblePluginAPI::class)
+    private fun transformDataFrameExpression(
+        expression: IrDeclarationReference,
+        ownerName: Name,
+        receiver: IrExpression?,
+        data: ContainingDeclarations
+    ): IrCall {
+        val alsoReference = pluginContext.referenceFunctions(FqName("kotlin.also")).single()
+
+        val result = IrCallImpl(-1, -1, expression.type, alsoReference, 1, 1).apply {
+            this.extensionReceiver = expression
+            putTypeArgument(0, expression.type)
+
+            val symbol = IrSimpleFunctionSymbolImpl()
+            val alsoLambda = IrFunctionImpl(
+                -1,
+                -1,
+                IrDeclarationOrigin.LOCAL_FUNCTION_FOR_LAMBDA,
+                symbol,
+                Name.special("<anonymous>"),
+                DescriptorVisibilities.LOCAL,
+                Modality.FINAL,
+                pluginContext.irBuiltIns.unitType,
+                isInline = false,
+                isExternal = false,
+                isTailrec = false,
+                isSuspend = false,
+                isOperator = false,
+                isInfix = false,
+                isExpect = false
+            ).apply {
+                valueParameters = buildList {
+                    add(IrValueParameterImpl(
+                        -1,
```

Comment:
I'd add names to all arguments, also above. This is risky when API changes occur plus difficult to review